src/python/grpcio/grpc/experimental/aio/_channel.py

```diff
@@ -383,22 +351,32 @@ def __init__(self, target: str, options: ChannelArgumentType,
         # No new calls will be accepted by the Cython channel.
         self._channel.closing()
 
-        if grace:
-            # pylint: disable=unused-variable
-            _, pending = await asyncio.wait(self._ongoing_calls.calls,
-                                            timeout=grace,
-                                            loop=self._loop)
-
-            if not pending:
-                return
-
-        # A new set is created acting as a shallow copy because
-        # when cancellation happens the calls are automatically
-        # removed from the originally set.
-        calls = WeakSet(data=self._ongoing_calls.calls)
+        # Iterate through running tasks
+        tasks = _all_tasks()
+        calls = []
+        call_tasks = []
+        for task in tasks:
+            stack = task.get_stack(limit=1)
+            if not stack:
+                continue
+
+            # Locate ones created by `aio.Call`.
+            frame = stack[0]
+            if 'self' in frame.f_locals:
+                if isinstance(frame.f_locals['self'], _base_call.Call):
+                    calls.append(frame.f_locals['self'])
+                    call_tasks.append(task)
+
+        # If needed, try to wait for them to finish.
+        # Call objects are not always awaitables.
+        if grace and call_tasks:
+            await asyncio.wait(call_tasks, timeout=grace, loop=self._loop)
```

Comment:
Race condition? Isn't it possible for `call_tasks` to change as we await here?

Reply:
The `call_tasks` can only be referenced within this function, and I think they should be free of race condition.